src/precice/impl/SolverInterfaceImpl.cpp

```diff
@@ -1172,6 +1179,280 @@ void SolverInterfaceImpl::writeScalarData(
+    auto &     gradientValuesInternal = data.gradientValues();
+    const auto vertexCount            = gradientValuesInternal.cols() / context.getDataDimensions();
+
+    //Check if the index and dimensions are valid
+    PRECICE_CHECK(valueIndex >= -1,
+                  "Invalid value index ({}) when writing gradient scalar data. Value index must be >= 0. "
+                  "Please check the value index for {}",
+                  valueIndex, data.getName());
+
+    PRECICE_CHECK(0 <= valueIndex && valueIndex < vertexCount,
+                  "Cannot write data \"{}\" to invalid vertex ID ({}). "
+                  "Please make sure you only use the results from calls to setMeshVertex/Vertices().",
+                  context.getDataName(), valueIndex);
+
+    PRECICE_CHECK(data.getDimensions() == 1,
+                  "You cannot call writeGradientScalarData on the vector data type \"{0}\". "
+                  "Use writeVectorGradientData or change the data type for \"{0}\" to scalar.",
+                  data.getName());
+
+    // Values are entered derived in the spatial dimensions (#rows = #spatial dimensions)
+    for (int i = 0; i < _dimensions; i++) {
+      gradientValuesInternal(i, valueIndex) = gradientValue[i];
+    }
+  }
+}
+
+void SolverInterfaceImpl::writeBlockScalarGradientData(
+    int           dataID,
+    int           size,
+    const int *   valueIndices,
+    const double *gradientValues,
+    bool          rowsFirst)
+{
+
+  // Asserts and checks
+  PRECICE_TRACE(dataID, size);
+  PRECICE_CHECK(_state != State::Finalized, "writeBlockScalarGradientData(...) cannot be called after finalize().");
+  PRECICE_REQUIRE_DATA_WRITE(dataID);
+  if (size == 0)
+    return;
+
+  if (isDataGradientRequired(dataID)) {
+
+    PRECICE_CHECK(valueIndices != nullptr, "writeBlockScalarGradientData() was called with valueIndices == nullptr");
+    PRECICE_CHECK(gradientValues != nullptr, "writeBlockScalarGradientData() was called with gradientValues == nullptr");
+
+    // Get the data
+    WriteDataContext &context = _accessor->writeDataContext(dataID);
+    PRECICE_ASSERT(context.providedData() != nullptr);
+    mesh::Data &data = *context.providedData();
+
+    PRECICE_CHECK(data.hasGradient(), "Data \"{}\" has no gradient values available! Please set the gradient flag to true under the data attribute in the configuration file.", data.getName())
+
+    PRECICE_CHECK(data.getDimensions() == 1,
+                  "You cannot call writeBlockScalarGradientData on the vector data type \"{}\". Use writeBlockVectorGradientData or change the data type for \"{}\" to scalar.",
+                  data.getName(), data.getName());
+
+    PRECICE_CHECK(data.getSpacialDimensions() == _dimensions,
+                  "The spatial dimensions of the data object and of the mesh don't match. Please make sure to initialize {} space dimensions for the scalar data type \"{}\".",
+                  _dimensions, context.getDataName());
```

Comment:
This one should be an assertion, right?
I don't see how a user could cause this.